core/src/main/scala/org/bykn/bosatsu/codegen/python/Code.scala

```diff
@@ -468,7 +470,27 @@ object Code {
     def simplify: Expression = Lambda(args, result.simplify)
   }
   case class Apply(fn: Expression, args: List[Expression]) extends Expression {
-    def simplify: Expression = Apply(fn.simplify, args.map(_.simplify))
+    def simplify: Expression = {
+      fn.simplify match {
+        case Lambda(largs, result) if largs.length == args.length => 
+          // if this is a lambda, but the args don't match, let
+          // the python error
+          val subMap = largs.iterator.zip(args).toMap
+          val subs = substitute(subMap, result)
```

Comment:
have to be careful here... what about a case like `(lambda x: lambda y: x + y)(y)` I don't think we are correctly handling this. To handle it correctly, we need to do something like return `lambda y1: y + y1`.

Add a test for this before merging.